src/ui/hash.js

```diff
@@ -67,19 +70,47 @@ class Hash {
         if (mapFeedback) {
             // new map feedback site has some constraints that don't allow
             // us to use the same hash format as we do for the Map hash option.
-            hash += `#/${lng}/${lat}/${zoom}`;
+            hash += `/${lng}/${lat}/${zoom}`;
         } else {
-            hash += `#${zoom}/${lat}/${lng}`;
+            hash += `${zoom}/${lat}/${lng}`;
         }
 
         if (bearing || pitch) hash += (`/${Math.round(bearing * 10) / 10}`);
         if (pitch) hash += (`/${Math.round(pitch)}`);
-        return hash;
+
+        if (this._hashName && !mapFeedback) {
+            let found = false;
+            const parts = window.location.hash.slice(1).split('&').map(part => {
+                const key = part.split('=')[0];
+                if (key === this._hashName) {
+                    found = true;
+                    return `${key}=${hash}`;
+                }
+                return part;
+            }).filter(a => a);
+            if (!found) {
+                parts.push(`${this._hashName || ''}=${hash}`);
+            }
+            return `#${parts.join('&')}`;
+        }
+
+        return `#${hash}`;
+    }
+
+    _getCurrentHash() {
+        const hash = window.location.hash.replace('#', '');
+        if (this._hashName) {
+            const keyval = hash.split('&').map(
+                part => part.split('=')
+            ).find(part => part[0] === this._hashName);
+            return (keyval ? keyval[1] || '' : '').split('/');
```

Comment:
is the `|| ''` part necessary?

Reply:
Yes, we must ensure there is a string value to be splitted, if you remove it, this test fails with error:
```
        window.location.hash = '#map&foo=bar';

        t.false(hash._onHashChange());
```

Reply:
If you really want this out, we can do this:
```diff
diff --git a/src/ui/hash.js b/src/ui/hash.js
index 056d07fbf..de7d30967 100644
--- a/src/ui/hash.js
+++ b/src/ui/hash.js
@@ -103,14 +103,14 @@ class Hash {
             const keyval = hash.split('&').map(
                 part => part.split('=')
             ).find(part => part[0] === this._hashName);
-            return (keyval ? keyval[1] || '' : '').split('/');
+            return keyval && keyval.length === 2 && keyval[1].split('/');
         }
         return hash.split('/');
     }
 
     _onHashChange() {
         const loc = this._getCurrentHash();
-        if (loc.length >= 3 && !loc.some(v => isNaN(+v))) {
+        if (loc && loc.length >= 3 && !loc.some(v => isNaN(v))) {
             this._map.jumpTo({
                 center: [+loc[2], +loc[1]],
                 zoom: +loc[0],
```

Your call

Reply:
OK, I think leave it as is